src/WebJobs.Extensions.DurableTask.Analyzers/Analyzers/Orchestrator/CancellationTokenAnalyzer.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class CancellationTokenAnalyzer
+    {
+        public const string DiagnosticId = "DF0111";
+
+        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.CancellationTokenAnalyzerTitle), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.CancellationTokenAnalyzerMessageFormat), Resources.ResourceManager, typeof(Resources));
+        private const string Category = SupportedCategories.Orchestrator;
+        public const DiagnosticSeverity Severity = DiagnosticSeverity.Warning;
+
+        public static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, Severity, isEnabledByDefault: true);
+
+        public static bool RegisterDiagnostic(CompilationAnalysisContext context, SyntaxNode method)
+        {
+            var diagnosedIssue = false;
+
+            foreach (SyntaxNode descendant in method.DescendantNodes())
+            {
+                if (descendant is ParameterSyntax parameter)
+                {
+                    var identifierType = parameter.Type;
+                    if (identifierType != null && identifierType.ToString() == "CancellationToken")
```

Comment:
This currently flags all CancellationTokens as parameters (on orchestrators or methods used within orchestrators defined in user code), but I think I want to limit this to just CancellationTokens as parameters on the actual orchestration trigger method. We use CancellationTokens on timers, at least.

Reply:
Right, we want to target just the CancellationTokens in the orchestrator trigger method signatures. Other uses of CancellationTokens should be fine.

Reply:
@ConnorMcMahon @cgillum I could also add another analyzer that flags cancellationTokenSource.Cancel() if its in a while loop.

Reply:
If the cancellationTokenSource variable declaration is not in the same while loop.

Reply:
I think there are valid cases where `CancellationTokenSource.Cancel()` can be safely used in a while loop. Is there a GitHub issue you can point us to that shows where this is problematic? I recall there was an issue where people used `WaitForExternalEvent` in a while loop that was problematic (and cancellation tokens were involved), but I can't remember what the exact details were. That might be appropriate for separate PR though.

Reply:
I don't actually have context on this, I just came across this comment and thought it might be helpful: https://github.com/Azure/azure-functions-durable-extension/issues/1633#issuecomment-756889845

Reply:
It may be worth solidifying in what exact cases it is unsafe to use cancellation tokens in an orchestrator function, and postpone this analyzer for a future release so it doesn't block the rest of this PR

Reply:
Just to reiterate, the only known unsafe usage of cancellation tokens is the one I mentioned earlier.
> we want to target just the CancellationTokens in the orchestrator trigger method signatures

Reply:
Correct. That is what this analyzer is doing. It only looks at methods with [OrchestrationTrigger] and only looks at ParameterSyntax.

Reply:
My apologies, I definitely was misreading this (and the original issue). I've reflected that in the issue discussion. I have no problem with that scope for this analyzer.